src/compiler/emitter.ts

```diff
@@ -4899,7 +4894,8 @@ namespace ts {
                 return node.text;
             }
 
-            return getSourceTextOfNodeFromSourceFile(currentSourceFile!, node, includeTrivia);
+            Debug.assertIsDefined(currentSourceFile);

```

Comment:
This will throw if you try to print an AST containing a mix of parse tree and non-parse tree nodes without providing a SourceFile. Since this function only gets called for Identifiers, PrivateIdentifiers, and literals (StringLiteral, NumericLiteral, etc.), if you add `|| !currentSourceFile` to the last `else if` branch above, then we probably won't hit this line when there's no source file available.

For example:
```ts
            else if (isLiteralExpression(node) && (nodeIsSynthesized(node) || !node.parent || !currentSourceFile)) {
                return node.text;
            }
```

Reply:
I can do that, but that won't actually "prove" that `currentSourceFile` is not undefined here; are you wanting the assert to be left in with the above change, or should I go back to a `currentSourceFile!`? (I'd prefer to not have any of those left that aren't the closure ones.)

Reply:
We'd probably need to rewrite the function to be more reliable from a "proof" perspective:

```ts
        function getTextOfNode(node: Identifier | PrivateIdentifier | LiteralExpression, includeTrivia?: boolean): string {
            if (isGeneratedIdentifier(node)) {
                return generateName(node);
            }
            if (isStringLiteral(node) && node.textSourceNode) {
                return getTextOfNode(node.textSourceNode, includeTrivia);
            }
            const sourceFile = currentSourceFile; // const needed for control flow
            const canUseSourceFile = !!sourceFile && !!node.parent && !nodeIsSynthesized(node);
            if (isMemberName(node)) {
                if (!canUseSourceFile || getSourceFileOfNode(node) !== getOriginalNode(sourceFile)) {
                    return idText(node);
                }
            }
            else {
                Debug.assertNode(node, isLiteralExpression);
                if (!canUseSourceFile) {
                    return node.text;
                }
            }
            return getSourceTextOfNodeFromSourceFile(sourceFile, node, includeTrivia);
        }
```
(plus a cast in `generateNameforNode` in the `SyntaxKind.Identifier` branch)

Reply:
Thanks, I applied that version.